src/PDOutline.jl

```diff
@@ -0,0 +1,407 @@
+## Data Structures
+"""
+```
+    PDOutlineItem
+```
+Representation of PDF document Outline item.
+
+It is currently defined as `Dict{Symbol, Any}`.
+Check description of function `pdDocGetOutline` for more information.
+
+Methods which operate on this structure:
+- `item_level(item::PDOutlineItem)` - return nesting level of item inside whole Outline.
+"""
+PDOutlineItem = Dict{Symbol, Any}
+
+"""
+```
+    PDOutline
+```
+Representation of PDF document Outline (Table of Contents).
+
+It is currently defined as `Vector{Union{PDOutlineItem, Vector}}`.
+Check description of function `pdDocGetOutline` for more information.
+
+Methods which operate on this structure:
+- `items_count(o::PDOutline; depth::Number = Inf)` - return number of items inside Outline.
+- `items(o::PDOutline; depth::Number = Inf)` - return iterator thru Outline items.
+"""
+PDOutline = Vector{Union{PDOutlineItem, Vector}}
+
+####################################################################################
+## Reading Outline from PDF
+function get_outline_node_compact(
+            cosDoc::CosDoc,
+            first_ref::CosIndirectObjectRef,
+            last_ref::CosIndirectObjectRef,
+            curr_depth::Int,
+            max_depth::Number,
+            index::Vector{Int})
+    arr = PDOutline()
+    support_index = length(index) > 0
+    curr_ref = first_ref
+    while true
+        obj = cosDocGetObject(cosDoc, curr_ref)
+        title_ref = get(obj, cn"Title")
+        title = cosDocGetObject(cosDoc, title_ref)
+        if title !== CosNull
+            item = PDOutlineItem()
+            title isa CosIndirectObject && (title = title.obj)
+            item[:Title] = convert(CDTextString, title)
+            if support_index
+                item[:Index] = tuple(index...)
+                index[end] = index[end] + 1
+            elseif curr_depth > 0
+                item[:Level] = curr_depth + 1
+            end
+            push!(arr, item)
+            ref_nest = get(obj, cn"First")
+            if ref_nest !== CosNull && curr_depth < max_depth
+                support_index && push!(index, 1)
+                push!(arr, get_outline_node_compact(cosDoc, ref_nest, get(obj, cn"Last"), curr_depth + 1, max_depth, index))
+                if support_index
+                    pop!(index)
+                    index[end] = index[end] + 1
+                end
+            end
+        end
+        curr_ref == last_ref && break
+        curr_ref = get(obj, cn"Next")
+        curr_ref === CosNull && break
+    end
+    return arr
+end
+
+get_outline_node_compact(
+            ::CosDoc,
+            ::CosNullType,
+            ::CosObject,
+            ::Int,
+            ::Number,
+            ::Vector{Int}) = nothing
+
+## Refers to PDF 32000-1:2008 / 12.4.2
+function find_label_for_page(cosDoc::CosDoc, values::Vector{Tuple{Int,CosObject}},
+                             pgnum::Int, ::String)
```

Comment:
Labels have no direct association with outlines from an API point of view. This functionality can be removed.

Reply:
Please see my comment above.
They are stored in different structure, but the page numbers are also stored in different structure.
Labels (if present in document) are being displayed on every outline view and printed on every Table of Contents.